platform/darwin/src/local_glyph_rasterizer.mm

```diff
@@ -124,52 +166,45 @@ CGContextHandle context(CGBitmapContextCreate(
     if (!context) {
         throw std::runtime_error("CGBitmapContextCreate failed");
     }
-
-    CFStringRef keys[] = { kCTFontAttributeName };
-    CFTypeRef values[] = { font };
-
-    CFDictionaryRefHandle attributes(
-        CFDictionaryCreate(kCFAllocatorDefault, (const void**)&keys,
-            (const void**)&values, sizeof(keys) / sizeof(keys[0]),
-            &kCFTypeDictionaryKeyCallBacks,
-            &kCFTypeDictionaryValueCallBacks));
-
-    CFAttributedStringRefHandle attrString(CFAttributedStringCreate(kCFAllocatorDefault, *string, *attributes));
-    CTLineRefHandle line(CTLineCreateWithAttributedString(*attrString));
+    
+    CFArrayRef glyphRuns = CTLineGetGlyphRuns(*line);
+    CTRunRef glyphRun = (CTRunRef)CFArrayGetValueAtIndex(glyphRuns, 0);
+    CFRange wholeRunRange = CFRangeMake(0, CTRunGetGlyphCount(glyphRun));
+    CGSize advances[CTRunGetGlyphCount(glyphRun)];
+    CTRunGetAdvances(glyphRun, wholeRunRange, advances);
+    metrics.advance = advances[0].width;
+    
+    CGFloat ascent;
+    CGFloat descent;
+    CGFloat leading;
+    CTRunGetTypographicBounds(glyphRun, wholeRunRange, &ascent, &descent, &leading);
     
     // Start drawing a little bit below the top of the bitmap
-    CGContextSetTextPosition(*context, 0.0, 5.0);
+    CGContextSetTextPosition(*context, 0.0, descent);
     CTLineDraw(*line, *context);
```

Comment:
This new code is pretty robust, even doing away with the hard-coded left, top, and advance metrics that were in the old code. But it falls short of the [coretext-hack](https://github.com/mapbox/mapbox-gl-native/compare/coretext-hack) branch (https://github.com/mapbox/mapbox-gl-native/issues/7862#issuecomment-362798522) in that individual characters are still being rasterized in isolation. If we were to enable local glyph rasterizing for all scripts today, that would cause some combining characters to be rendered with unsightly spacing [dotted circles](https://en.wikipedia.org/wiki/Dotted_circle) (◌) underneath.

coretext-hack didn’t have these artifacts. I think it’s because `CTFontDrawGlyphs()` already knows the glyphs will be reassembled into a line, whereas `CTLineDraw()` assumes the glyphs have already been reassembled. I think replacing `CTLineDraw()` with `CTFontDrawGlyphs()` would do the trick. However, making `GlyphManager::generateLocalSDF()` operate on an entire string, not individual glyphs, would not only address the dotted circle problem but weaken our incorrect reliance on a one-to-one mapping between characters and glyphs.

Reply:
Using `CTFontDrawGlyphs()` resulted in the same rendered output, including the dotted circles. It’s definitely because we’re rendering one character at a time.

The coretext-hack branch’s approach does require rendering one glyph at a time using `CTFontDrawGlyphs()`, but `getGlyphDependencies()` calculates those glyph IDs based on an entire run of text ahead of time. In fact, that branch redefines `GlyphID` to pair the numeric glyph ID with a font name, since a given codepoint can be represented by a different glyph depending on the font.

Essentially, we need this code that currently conflates codepoints with glyphs to ask a platform-specific implementation to perform shaping and map these codepoints to actual glyph IDs, and the font name is important for preventing collisions between different glyphs in different fonts that happen to have the same ID:

https://github.com/mapbox/mapbox-gl-native/blob/66ac55445998718b28a6fbd4a4094a10426627a9/src/mbgl/layout/symbol_layout.cpp#L179-L182

I can’t think of a less invasive way to solve the combining character issue, because shaping happens so far from glyph rendering in mbgl’s current architecture. The font name could be optional for platforms that don’t perform shaping, but the diff will still be large. Clearly, that refactoring will need to take place in a separate PR.